frontend/catalyst/qfunc.py

```diff
@@ -106,3 +121,118 @@ def _eval_quantum(*args):
         args_flat = tree_flatten(args)[0]
         res_flat = func_p.bind(flattened_fun, *args_flat, fn=self)
         return tree_unflatten(out_tree_promise(), res_flat)[0]
+
+
+# pylint: disable=protected-access,no-member,not-callable
+def dynamic_one_shot(qnode):
+    """Transform a QNode to into several one-shot tapes to support dynamic circuit execution.
+
+    Args:
+        qnode (QNode): a quantum circuit which will run n-shot times
+
+    Returns:
+        qnode (QNode):
+
+        The transformed circuit to be run n-shot times such as to simulate dynamic execution.
+
+
+    **Example**
+
+    Consider the following circuit:
+
+    .. code-block:: python
+
+        dev = qml.device("lightning.qubit", shots=100)
+        params = np.pi / 4 * np.ones(2)
+
+        @qjit
+        @dynamic_one_shot
```

Comment:
Follow up, can the decorator be applied both before and after the `qjit` decorator?

Reply:
I think Vincent is just following the established style here, where all functions from pennylane are used via `qml.`, while Catalyst functions are assumed imported as `from catalyst import *`.
Changing this would probably be a separate decision I think.

Reply:
I would be all for using QNode options exclusively. This would simplify the code and the UI.

Reply:
> Follow up, can the decorator be applied both before and after the qjit decorator?

Missed this question earlier but I would think not. The decorator has to be applied to a qnode, and the qjit object is not a qnode.

I agree with Vincent that only exposing the qnode option is simpler and allows us more flexibility to change things under the hood rather than having to maintain an "official transform".

Reply:
Should we merge this PR as is and fix this in the PR introducing `mcm_method`?

Reply:
Definitely, not a blocker for this one :) Although in that case not importing it top-level might useful so we don't accidentally forget it there 😅

Reply:
Do you mean we should decorate with `catalyst.dynamic_one_shot` instead?

Reply:
It lives in `catalyst.qfunc` right? So in that case we wouldn't import `dynmic_one_shot` in the `catalyst.__init__`, and in the tests invoke/import as `catalyst.qfunc.dynamic_one_shot`. 

The docstring is fine as is since it refers to itself.

Reply:
I see, I'll resolve it this way. Thanks.

Reply:
Thanks, that makes sense. Resolving.